### DISCUSSION TITLE
HTTPS with Nginx Possible?

### DISCUSSION THREAD

**Discussion by jlpoolen** [maplibre/martin#1398] [Q&A] [answered]
I have an instance of martin running on a Debian (v. 12 "bookworm") server:
```
jlpoole@debian1:~/martin/rust/martin$ git --no-pager log -1
commit 68aca37ba216e3b2f98b1a6a57743568ae36378a (HEAD -> main, origin/main, origin/HEAD)
Author: Yuri Astrakhan <YuriAstrakhan@gmail.com>
Date:   Thu Jun 13 18:33:29 2024 -0400
```

If I set up an HTML page and reference a JSON style sheet and the JSON style sheet specifies "**localhost**" instead of the server name, i.e. "debian1", then I can, on the same Debian server in an instance of Firefox, view a rendered map using HTTPS.  I cannot, however, access the page from another machine -- I get a CORS error.  Moreover, if I alter the **url** value in the style sheet to the server name in place of "localhost", then I receive a CORS error in Firefox on the Debian server.

Question: Is it possible to run nginx on a server in front of the martin tile server and be able to access the map HTML page via HTTPS from another server?

Here are my notes:

Steps to make nginx.conf work after downloading from GitHub at:
  https://raw.githubusercontent.com/maplibre/martin/main/demo/frontend/nginx.conf

Start up attempt 1) no changes to nginx.conf
start-up errors out:

```
Jul 03 08:03:18 debian1 nginx[1182238]: 2024/07/03 08:03:18 [emerg] 1182238#1182238: host not found in upstream "tiles:3000" in /etc/nginx/nginx.conf:46
Jul 03 08:03:18 debian1 nginx[1182238]: nginx: configuration file /etc/nginx/nginx.conf test failed
Jul 03 08:03:18 debian1 systemd[1]: nginx.service: Control process exited, code=exited, status=1/FAILURE
```

Change 1
```
    upstream tiles_upstream {
        #server tiles:3000;
	server debian1:3000;
    }
```
    
start up attempt 2) success on same server, but fail on remote server.
    
Access Attempts 1:  
    A) on debian1 with Firefox: https://debian1/test/martin_map_SSL_2.html#14.12/44.9384/-123.02344 **works**
    B) remote laptop with Firefox:  https://debian1/test/martin_map_SSL_2.html#14.12/44.9384/-123.02344 **fails** with:
    
     `Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost:3000/jlptiles. (Reason: CORS request did not succeed). Status code: (null).`
    
Status of  nginx configuration file:
    
	root@debian1:/etc/nginx# diff nginx.conf.GitHub nginx.conf.GitHub.MINIMAL_MODIFICATIONS
	46c46,47
	<         server tiles:3000;
	---
	>         #server tiles:3000;
	>       server debian1:3000;
	root@debian1:/etc/nginx# date; diff nginx.conf.GitHub nginx.conf.GitHub.MINIMAL_MODIFICATIONS
	Wed Jul  3 09:05:55 AM PDT 2024
	46c46,47
	<         server tiles:3000;
	---
	>         #server tiles:3000;
	>       server debian1:3000;
	root@debian1:/etc/nginx# 

martin_map_SSL_2.html portions:
```
    ...
    15       <link rel="stylesheet" type="text/css" href="https://debian1/test/maplibre-gl.css"> 
    16       <script src="https://debian1/test/maplibre-gl.js"></script>
    ...
    44    <script>
    45      var map = new maplibregl.Map({
    46          container: 'map',
    47          style: 'https://debian1/test/martin_style_SSL_debian1.json',
    48          hash: true,
    49          center: [-123.02121, 44.935], // starting position [lng, lat] Salem, OR
    50          zoom: 16, // starting zoom
    51          maplibreLogo: true
    52      });
    53      map.scrollZoom.enable();
    54   </script>
    ...
```
JSON, martin_style_SSL_debian1.json, style:
    
```
    jlpoole@debian1:/usr/share/nginx/html/test$ head -n 27 martin_style_SSL_debian1.json|nl
     1  {
     2    "version": 8,
     3    "name": "Test MapLibre Martin Style",
     4    "metadata": {"maputnik:renderer": "mlgljs"},
     5    "center": [-122.98, 44.95],
     6    "zoom": 14,
     7    "bearing": 0,
     8    "pitch": 0,
     9    "sources": {
    10      "jlptiles": {
    11        "type": "vector",
    12        "scheme": "xyz",
    13        "url": "http://localhost:3000/jlptiles"
    14      },
    15      "attribution": {
    16        "attribution": "<a href=\"https://www.openmaptiles.org/\" target=\"_blank\">&copy; OpenMapTiles</a> <a href=\"https://www.openstreetmap.org/copyright\" target=\"_blank\">&copy; OpenStreetMap contributors</a>",
    17        "type": "vector"
    18      }
    19    },
    20    "sprite": [
    21      {
    22        "id": "default",
    23        "url": "https://debian1/test/sprite_dir/sprite"
    24      }
    25    ],
    26   "glyphs":"https://debian1/test/fontspbf/{fontstack}/{range}.pbf",
    27    "layers": [
    ...
jlpoole@debian1:/usr/share/nginx/html/test$ 
```
martin_3.config:
```
    mbtiles:
  sources: 
    jlptiles: ./tiles.mbtiles
fonts: ./fonts
#listen_addresses: 'debian1:3000'
```
    
martin server console:
    
```
    jlpoole@debian1:/usr/share/nginx/html/test$ ./martin -c martin_3.config 
[2024-07-03T00:54:09Z INFO  martin] Starting Martin v0.13.0
[2024-07-03T00:54:09Z INFO  martin] Using martin_3.config
[2024-07-03T00:54:09Z INFO  martin::config] Initializing main cache with maximum size 536870912B
[2024-07-03T00:54:09Z INFO  martin::file_config] Configured source jlptiles from /usr/share/nginx/html/test/tiles.mbtiles
[2024-07-03T00:54:09Z INFO  martin::fonts] Configured font Noto Sans Bold Italic with 2840 glyphs (0000-FFFD) from ./fonts/NotoSans-BoldItalic.ttf
[2024-07-03T00:54:09Z INFO  martin::fonts] Configured font Noto Sans Italic with 2840 glyphs (0000-FFFD) from ./fonts/NotoSans-Italic.ttf
[2024-07-03T00:54:09Z INFO  martin::fonts] Configured font Noto Sans Regular with 2840 glyphs (0000-FFFD) from ./fonts/NotoSans-Regular.ttf
[2024-07-03T00:54:09Z INFO  martin::fonts] Configured font Noto Sans Bold with 2840 glyphs (0000-FFFD) from ./fonts/NotoSans-Bold.ttf
[2024-07-03T00:54:09Z WARN  martin::fonts] Ignoring duplicate font Noto Sans Regular from ./fonts/Noto Sans Regular.ttf because it was already configured from ./fonts/NotoSans-Regular.ttf
[2024-07-03T00:54:09Z INFO  martin] Use --save-config to save or print Martin configuration.
[2024-07-03T00:54:09Z INFO  martin] Martin has been started on 0.0.0.0:3000.
[2024-07-03T00:54:09Z INFO  martin] Use http://0.0.0.0:3000/catalog to get the list of available sources.

```
Also, I have HOST defined as "debian1" in the console where the martin server is started:

```
	jpoole@debian1:/usr/share/nginx/html/test$ echo $HOST
	debian1
	jlpoole@debian1:/usr/share/nginx/html/test$ 
```
	
If I change "localhost" to "debian1" for the **url** value in the JSON,
```
	jlpoole@debian1:/usr/share/nginx/html/test$ date;head -n 33  martin_style_SSL_debian1.json|nl|grep url
	Wed Jul  3 10:43:20 AM PDT 2024
	    13        "url": "http://debian1:3000/jlptiles"
	    23        "url": "https://debian1/test/sprite_dir/sprite"
	jlpoole@debian1:/usr/share/nginx/html/test$ 

```
then in Firefox running on debian1 where I previously succeeded in viewing a rendered map, I get a CORS error:
	`Blocked loading mixed active content "http://debian1:3000/jlptiles"
`
So, whether a map renders or not is dependent upon the value of the server name in the JSON **url** specification.

I tried with the JSON "url" set to "debian1" these two variations of /etc/hosts, and still I get a CORS error:

```
	jlpoole@debian1:/usr/share/nginx/html/test$ cat -n /etc/hosts
	     1  #127.0.0.1      localhost
	     2  #127.0.0.1      debian1.salemdata.net   debian1 debian1.salemdata.us
	   ...
	    17  192.168.1.142 localhost debian1 debian1.salemdata.us
	    18
	jlpoole@debian1:/usr/share/nginx/html/test$ sudo nano /etc/hosts
	jlpoole@debian1:/usr/share/nginx/html/test$ cat -n /etc/hosts
	     1  127.0.0.1       localhost
	     2  127.0.0.1       debian1.salemdata.net   debian1 debian1.salemdata.us
	 ..
	    17  192.168.1.142 localhost debian1 debian1.salemdata.us
	    18
	jlpoole@debian1:/usr/share/nginx/html/test$ 

```
What am I missing or what configuration setting is awry?

**Comment by jlpoolen** [comment]
Also, the version of the JavaScript & css is from:

```
jlpoole@debian1:/usr/local/src/maplibre-gl-js$ date;git --no-pager log -1
Wed Jul  3 12:00:04 PM PDT 2024
commit f969cce96bdecfb90389e1cde5b6a5614182c4b5 (HEAD -> main, origin/main, origin/HEAD)
Author: Bart Louwers <bart.louwers@gmail.com>
Date:   Fri Jun 21 12:28:38 2024 +0200
```

**Comment by nyurik** [comment]
i think there are a few unrelated issues here:
* you can setup nginx to work as you entrypoint proxy - all requests should go to your `nginxserver`, and nginx would forward some of them to the tile server. See https://maplibre.org/martin/run-with-nginx.html for some ideas. You do NOT have to use docker compose - the nginx config would still be fairly similar.  Having this setup will not require any CORS because all requests will appear as coming from the same server (from the browser's perspective)
* you can set up https support with nginx following the standard nginx configuration - there is an example of how to do that in the `/demo` dir, but nginx docs would probably tell you more

**Reply by jlpoolen** [reply]
Thank you @nyurik.  As far as I know, I am not running docker at all (see output below).  

I did build the openmaptiles project using which resulted in two dockers containers built by the quickstart.sh, but I have halted those docker instances and harvested the components, mainly, the mbtiles file. I do not want to use docker in my evenutal set-up.  The console results below of "docker ps" show that there are no instances of docker running.  I copied the mbtiles database into /usr/share/nginx/html/test so I could wrap up everything into a zip archive in case someone wanted a complete universe of what I had.  I also staged two types of fonts in this test area.

```
jlpoole@debian1:/usr/share/nginx/html/test$ date; sudo docker ps
Wed Jul  3 03:13:33 PM PDT 2024
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
jlpoole@debian1:/usr/share/nginx/html/test$ 
```
Further, here's my testing area:

```
jlpoole@debian1:/usr/share/nginx/html/test$ date;ls -lath
Wed Jul  3 03:15:40 PM PDT 2024
total 370M
drwxr-xr-x 5 jlpoole jlpoole 4.0K Jul  3 10:39 .
-rw-r--r-- 1 jlpoole jlpoole 148K Jul  3 10:39 martin_style_SSL_debian1.json
-rwxr-xr-x 1 jlpoole jlpoole  100 Jul  2 17:54 martin_3.config
lrwxrwxrwx 1 jlpoole jlpoole   18 Jul  2 15:21 maplibre-gl.js -> maplibre-gl-dev.js
-rw-r--r-- 1 jlpoole jlpoole 776K Jul  2 15:21 maplibre-gl.js_PRODUCTION
-rw-r--r-- 1 jlpoole jlpoole 2.3M Jul  2 15:20 maplibre-gl-dev.js
-rw-r--r-- 1 jlpoole jlpoole 1.8K Jul  2 13:13 martin_map_SSL_2.html
drwxr-xr-x 5 jlpoole jlpoole 4.0K Jul  2 08:45 fontspbf
drwxr-xr-x 2 jlpoole jlpoole 4.0K Jul  2 08:37 fonts
-rw-r--r-- 1 jlpoole jlpoole 1.3K Jul  2 07:07 martin_map.html
-rw-r--r-- 1 jlpoole jlpoole 148K Jul  2 07:07 martin_style_debian1.json
-rw-r--r-- 1 jlpoole jlpoole 148K Jul  1 20:18 martin_style_SSL.json
-rw-r--r-- 1 jlpoole jlpoole 1.7K Jul  1 20:14 martin_map_SSL.html
-rwxr-xr-x 1 jlpoole jlpoole  29M Jul  1 20:01 martin
-rw-r--r-- 1 jlpoole jlpoole  64K Jul  1 19:57 maplibre-gl.css
-rw-r--r-- 1 root    root      63 Jul  1 19:29 index.html
-rwxr-xr-x 1 jlpoole jlpoole 337M Jul  1 19:05 tiles.mbtiles
-rw-r--r-- 1 jlpoole jlpoole 148K Jul  1 19:02 martin_style.json
drwxr-xr-x 2 jlpoole jlpoole 4.0K Jul  1 18:53 sprite_dir
drwxrwxrwx 3 root    root    4.0K Jul  1 18:45 ..
jlpoole@debian1:/usr/share/nginx/html/test$ 
```
I'm using nginx exclusively, at the moment.  (I have been trying to get Apache to work as the "entrypoint proxy", but was not successful, so I stepped back and am using nginx as the sole server to accomplish the goal of having nginx respond to HTTPS requests from a browser and then querying the martin server with HTTP and forwarding the results back to the requestor via HTTPS.)  

I did have to alter the JSON style file, fixing two nonconforming structures under the roads features.  I used Maputnik to accomplish that goal and make the martin_style_SSL.json valid. 

I am trying to use only: 1) nginx and 2) martin.

I included the nginx configuration in my initial posting here which was copied and slightly altered as shown.

**Reply by nyurik** [reply]
lets not discuss SSL - this part can be done later. You do NOT need docker or docker compose.  Your nginx.conf can forward requests to Martin backend - see the link i gave above - it shows the conf file that is similar to what you need, either in or outside of the docker

**Reply by jlpoolen** [reply]
Yes, I have successfully combined nginx and martin using HTTP only.  No problems with HTTP.  I accomplished that several days ago.  Now I'm trying to address SSL, and as you can see, I can accomplish that if my browser is on the same server as the nginx and martin.  But if my browser is on another machine, I run into CORS issues.

**Reply by jlpoolen** [reply]
To demonstrate the HTTP Port 80 works (both with on-server browser and off-server browser) vs. my attempt for HTTPS Port 443, I should have included a listing of my /etc/nginx directory showing how I use soft links to switch to the various configurations:

```
root@debian1:/etc/nginx# ls -la
total 112
drwxrwxrwx   8 root    root     4096 Jul  3 08:08 .
drwxrwxrwx 139 root    root    12288 Jul  3 10:44 ..
drwxr-xr-x   2 root    root     4096 Mar 14  2023 conf.d
-rw-r--r--   1 root    root     1125 Mar 14  2023 fastcgi.conf
-rw-r--r--   1 root    root     1055 Mar 14  2023 fastcgi_params
-rw-r--r--   1 root    root     2837 Mar 14  2023 koi-utf
-rw-r--r--   1 root    root     2223 Mar 14  2023 koi-win
-rw-r--r--   1 root    root     4338 Mar 14  2023 mime.types
drwxr-xr-x   2 root    root     4096 Mar 14  2023 modules-available
drwxr-xr-x   2 root    root     4096 Jun 27 07:59 modules-enabled
lrwxrwxrwx   1 root    root       39 Jul  3 08:07 nginx.conf -> nginx.conf.GitHub.MINIMAL_MODIFICATIONS
-rw-r--r--   1 jlpoole jlpoole  1537 Jul  2 16:51 nginx.conf.ChatGPT
-rw-r--r--   1 jlpoole jlpoole  2950 Jun 30 14:56 nginx.conf.GitHub
-rw-r--r--   1 jlpoole jlpoole  2973 Jul  3 08:08 nginx.conf.GitHub.MINIMAL_MODIFICATIONS
-rwxrwxrwx   1 root    root     6059 Jul  2 16:33 nginx.conf.MARTIN
-r--r--r--   1 root    root     1446 Mar 14  2023 nginx.conf.ORIGINAL
-rw-r--r--   1 jlpoole jlpoole  2354 Jul  2 07:02 nginx.conf.PORT_80_debian_WORKS
-r-xr-xr-x   1 root    root     1684 Jul  1 20:06 nginx.conf.PORT_80_WORKS
-rw-r--r--   1 root    root      180 Mar 14  2023 proxy_params
-rw-r--r--   1 root    root      636 Mar 14  2023 scgi_params
drwxr-xr-x   2 root    root     4096 Jun 27 07:59 sites-available
drwxr-xr-x   2 root    root     4096 Jun 27 07:59 sites-enabled
drwxr-xr-x   2 root    root     4096 Jun 27 07:59 snippets
-rw-r--r--   1 root    root      664 Mar 14  2023 uwsgi_params
-rw-r--r--   1 root    root     3071 Mar 14  2023 win-utf
root@debian1:/etc/nginx# 

```

**Reply by jlpoolen** [reply]
To demonstrate that the nginx->martin server is working, here are screen shots of Firefox accessing using HTTP martin_map.html on the server where the nginx & martin are running as well as a remote Windows machine:

![tvnviewer_2024-07-03_15-46-54](https://github.com/maplibre/martin/assets/227313/a893a8b2-feb5-4381-9fe2-5e372fff2d78)
![2024-07-03_15-48-45](https://github.com/maplibre/martin/assets/227313/31dac703-2029-47f5-9647-6a03977c70c1)

Note: nginx is running this configuration:
```
root@debian1:/etc/nginx# ls -la nginx.conf
lrwxrwxrwx 1 root root 31 Jul  3 15:45 nginx.conf -> nginx.conf.PORT_80_debian_WORKS
root@debian1:/etc/nginx# 
```
Here is the heart of the HTML for test/martin_map.html:
```
  <h1>Martin Server port 3000 Via HTTP Test 1</h1>
  <div id="map"></div>
  <script>
    var map = new maplibregl.Map({
	container: 'map',
	style: 'http://debian1/test/martin_style_debian1.json',
	hash: true,
	center: [-123.02121, 44.935], // starting position [lng, lat] Salem, OR
        zoom: 16, // starting zoom
        maplibreLogo: true
    });
    map.scrollZoom.enable();
```

and I have the referenced JSON style sheet "url" value set at:
```
jlpoole@debian1:/usr/share/nginx/html/test$ date;head -n 14 martin_style_debian1.json |nl|tail -n 6
Wed Jul  3 03:56:45 PM PDT 2024
     9    "sources": {
    10      "jlptiles": {
    11        "type": "vector",
    12        "scheme": "xyz",
    13        "url": "http://debian1:3000/jlptiles"
    14      },
jlpoole@debian1:/usr/share/nginx/html/test$ 
```
**Edit** 7/3/24 4:30 PM PDT: I meant to include proof of what configuration nginx was using and do so now:
```
root@debian1:/etc/nginx# unlink nginx.conf
root@debian1:/etc/nginx# ln -s nginx.conf.PORT_80_debian_WORKS nginx.conf
root@debian1:/etc/nginx# systemctl start nginx
root@debian1:/etc/nginx# ls -la nginx.conf
lrwxrwxrwx 1 root root 31 Jul  3 15:45 nginx.conf -> nginx.conf.PORT_80_debian_WORKS
root@debian1:/etc/nginx# 
```

**Reply by nyurik** [reply]
@jlpoolen when you browse in a non-TLS mode, could you look at the network requests - see what connections are actually being made?  I suspect you do not use the nginx proxy for tiles (i might be wrong though)

**Reply by jlpoolen** [reply]
@nyurik  **You have identified the problem.**  The JSON style sheet, martin_style_debian1.json, for the port 80 test has is url value set to port 3000 so nginx is completely out of the circuit when using that url.
```
jlpoole@debian1:/usr/share/nginx/html/test$ date;head -n 14 martin_style_debian1.json |nl|tail -n 6
Thu Jul  4 06:45:59 AM PDT 2024
     9    "sources": {
    10      "jlptiles": {
    11        "type": "vector",
    12        "scheme": "xyz",
    13        "url": "http://debian1:3000/jlptiles"
    14      },
jlpoole@debian1:/usr/share/nginx/html/test$
```
So specifying "http://debian1:3000" would necessarily bypass the nginx so no proxying would occur, the request would instead go directly to martin.  The martin server does accept HTTP requests so for HTTP tests, it responds.  I see (below) that the JSON style sheet I used for the HTTPS tests also has HTTP, not HTTPS as the protocol.  So you suspicion about not being proxied is correct.  The error of my configuration is in the JSON style sheet which should point to port 80 and let nginx use the match of "jlptiles" to redirect (proxy) to the martin service on port 3000.

```
jlpoole@debian1:/usr/share/nginx/html/test$ date;head -n 14 martin_style_SSL_debian1.json |nl|tail -n 6
Thu Jul  4 06:55:31 AM PDT 2024
     9    "sources": {
    10      "jlptiles": {
    11        "type": "vector",
    12        "scheme": "xyz",
    13        "url": "http://localhost:3000/jlptiles"
    14      },
jlpoole@debian1:/usr/share/nginx/html/test$ 
```
_Thank you!  I_'ll tinker with the configurations and have the JSON style sheet point to the nginx server (for proxying) instead.  I'll update this Issue when I have some results to report.

**Reply by nyurik** [reply]
Don't change config. Your proxy should set several headers when forwarding to to martin, and martin will update the json. See nginx martin book I gave above

**Reply by jlpoolen** [reply, marked as answer]
The problem I encountered, CORS errors, has been resolved.  The problem was that my JSON style file specified a URL for the tiles to point directly to the martin server, when, it should have directed the inquiry through nginx using HTTPS.

I modified my JSON stylesheet for the HTTPS test to:

```
jlpoole@debian1:/usr/share/nginx/html/test$ date;head -n 14 martin_style_debian1.json |nl|tail -n 6
Thu Jul  4 07:35:22 AM PDT 2024
     9    "sources": {
    10      "jlptiles": {
    11        "type": "vector",
    12        "scheme": "xyz",
    13        "url": "http://debian1/jlptiles/"
    14      },
jlpoole@debian1:/usr/share/nginx/html/test$ 

```
And when I now make an HTTPS call to the testing page, i.e. https://debian1/test/martin_map_SSL_2.html#14.12/44.94176/-123.02251, I no longer receive CORS error messages and the call for "jlptiles" is given a status of "OK".  A map does not render. but I have learned thus far that unless everything in the style sheet is error free, there is a likelihood the map will not render and the error messages in a standard configuration are not identifying problems.  I will need to activate some debugging both on the martin server to see what comes in and goes out and then in the MapLibre-gl script (I have the development version which is not minified so I can add some debugging statements).  For posterity, here is my Firefox Console from a browser on a remote machine:
![2024-07-04_07-42](https://github.com/maplibre/martin/assets/227313/9d44cabb-5f11-49e6-bfe3-893ceabcf2f6)

Thank you, again, @nyurik
